### ISSUE TITLE
Errors trying to mount paths between qubes

### ISSUE THREAD

**Issue by bushchannelstepper** [Rudd-O/qubes-shared-folders#9]
Hi,

Built rpms for diod and this package on Qubes 4.2RC1.

Built and installed ok.  Got this error when trying to mount in an AppVM from a zpool on a drive attached to another standalone VM (fedora 38)

`[user@av ~]$ qvm-mount-folder storage /home/user content
mount: /home/user/content: cannot mount qvm://storage/home/user read-only.
       dmesg(1) may have more information after failed mount system call.
[user@av ~]$ `

I had to install sssd in dom0 to add pam_sss.so which was missing as I could see from journalctl there.  No further error visible in dom0, yet the error in the VM persists.

Trying to mount /home/user from a different AppVM to test, I get 

`[user@av ~]$ qvm-mount-folder personal /home/user content
error: unknown exit status 127`

Any idea what's going on here?

Also, would I be right in supposing there is no way to share ZFS filesystems within dom0?

Thanks

**Comment by Rudd-O** [comment]
Not really.  Need journal logs from all involved VMs and dom0.

**Comment by Rudd-O** [comment]
Did you try the latest release?  Some issues are fixed.

**Comment by Cyrixo** [comment]
Hi,
Same error, qubes 4.2.
```
dmesg
[ 1658.180710] audit: type=1101 audit(1710682107.233:196): pid=1618 uid=1000 auid=1000 ses=3 subj=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 msg='op=PAM:accounting grantors=pam_unix,pam_localuser acct="user" exe="/usr/bin/sudo" hostname=? addr=? terminal=/dev/pts/0 res=success'
[ 1658.181935] audit: type=1123 audit(1710682107.234:197): pid=1618 uid=1000 auid=1000 ses=3 subj=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 msg='cwd="/home/user" cmd="dmesg" exe="/usr/bin/sudo" terminal=pts/0 res=success'
[ 1658.183855] audit: type=1110 audit(1710682107.236:198): pid=1618 uid=1000 auid=1000 ses=3 subj=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 msg='op=PAM:setcred grantors=pam_env,pam_localuser,pam_unix acct="root" exe="/usr/bin/sudo" hostname=? addr=? terminal=/dev/pts/0 res=success'
[ 1658.192679] audit: type=1105 audit(1710682107.245:199): pid=1618 uid=1000 auid=1000 ses=3 subj=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 msg='op=PAM:session_open grantors=pam_keyinit,pam_limits,pam_keyinit,pam_limits,pam_systemd,pam_unix acct="root" exe="/usr/bin/sudo" hostname=? addr=? terminal=/dev/pts/0 res=success'
[ 1658.194904] audit: type=2300 audit(1710682107.247:200): pid=1619 uid=1000 auid=1000 ses=3 subj=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 msg='newrole: old-context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 new-context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 exe="/usr/bin/sudo" hostname=? addr=? terminal=/dev/pts/1 res=success'
```
I think the problem comes from selinux which is active with fedora-39.
If I disable selinux in the client with ```` sudo setenforce 0 ```` qvm-mount-folder run without error

**Comment by Atrate** [comment]
I can confirm the same issue on a Fedora 39 template.

Disabling SELinux makes it work again, re-enabling it makes it break (including already-mounted shares).

**Comment by Rudd-O** [comment]
I don't have a system with SELinux setup to test this, but the logs you pasted are not useful because they do not show any SELinux AVC denials.

Please reopen after you have logs I can use.

**Comment by kalbasit** [comment]
@Rudd-O I'm also seeing the same thing on a fresh Qubes 4.2 installation. It seems to come with SELinux enabled by default! Unfortunately, I don't see logs that can help you debug this, I put logs that I could extract in this Gist: https://gist.github.com/kalbasit/0b071467ac60bca2fd06acb2e98e086a

**Comment by kalbasit** [comment]
I added `/usr/bin/strace` right above this line, and I got a strace for it working and not working.
https://github.com/Rudd-O/qubes-shared-folders/blob/2c697ea5aaf90d962f6b61e677a7b2d4cf0e27cc/py/sharedfolders/programs.py#L317

When it's not working, I see:

```
access("/run/mount", R_OK|W_OK)         = 0                                                                                                                              
fsconfig(3, FSCONFIG_SET_STRING, "source", "qvm://keybase/home/user/keybase_"..., 0) = 0                               
fsconfig(3, FSCONFIG_SET_STRING, "trans", "fd", 0) = 0                                                                                                                   
fsconfig(3, FSCONFIG_SET_STRING, "rfdno", "0", 0) = 0                                                                                                                    
fsconfig(3, FSCONFIG_SET_STRING, "wfdno", "1", 0) = 0                                                                                                                    
fsconfig(3, FSCONFIG_SET_STRING, "version", "9p2000.L", 0) = 0
fsconfig(3, FSCONFIG_SET_STRING, "dfltuid", "2000", 0) = 0                                                                                                               fsconfig(3, FSCONFIG_SET_STRING, "dfltgid", "2000", 0) = 0
fsconfig(3, FSCONFIG_SET_STRING, "uname", "user", 0) = 0
fsconfig(3, FSCONFIG_SET_STRING, "aname", "/home/user/keybase_git/secrets", 0) = 0
fsconfig(3, FSCONFIG_CMD_CREATE, NULL, NULL, 0) = -1 EACCES (Permission denied)
```

When it's working, I see this

```
access("/run/mount", R_OK|W_OK)         = 0
fsconfig(3, FSCONFIG_SET_STRING, "source", "qvm://keybase/home/user/keybase_"..., 0) = 0
fsconfig(3, FSCONFIG_SET_STRING, "trans", "fd", 0) = 0
fsconfig(3, FSCONFIG_SET_STRING, "rfdno", "0", 0) = 0
fsconfig(3, FSCONFIG_SET_STRING, "wfdno", "1", 0) = 0
fsconfig(3, FSCONFIG_SET_STRING, "version", "9p2000.L", 0) = 0
fsconfig(3, FSCONFIG_SET_STRING, "dfltuid", "2000", 0) = 0
fsconfig(3, FSCONFIG_SET_STRING, "dfltgid", "2000", 0) = 0
fsconfig(3, FSCONFIG_SET_STRING, "uname", "user", 0) = 0
fsconfig(3, FSCONFIG_SET_STRING, "aname", "/home/user/keybase_git/secrets", 0) = 0
fsconfig(3, FSCONFIG_CMD_CREATE, NULL, NULL, 0) = 0
```

I put the entire strace logs in [this gist](https://gist.github.com/kalbasit/0c44a99167b7a7abd3eb7efc6ba4b0ff).

**Comment by kalbasit** [comment]
Running `sudo ausearch -m avc,user_avc,selinux_err -i` does not reveal anything useful but after making SELinux verbose with `semodule -DB` it does return something:

```
type=AVC msg=audit(12/02/2024 12:06:24.285:319) : avc:  denied  { write } for  pid=3051 comm=kworker/0:0 path=pipe:[33402] dev="pipefs" ino=33402 scontext=system_u:system_r:kernel_t:s0 tcontext=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0 
----
type=AVC msg=audit(12/02/2024 12:06:24.285:320) : avc:  denied  { write } for  pid=3051 comm=kworker/0:0 path=pipe:[33402] dev="pipefs" ino=33402 scontext=system_u:system_r:kernel_t:s0 tcontext=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 tclass=fifo_file permissive=0 
```

**Comment by kalbasit** [comment]
@bushchannelstepper @Rudd-O now that I added some helpful logs and a possible solution, can you please re-open the issue?

**Comment by Atrate** [comment]
> 
> ```cil
> module fix-qvm-mount-folder 1.0;
> 
> require {
>         type unconfined_t;
>         type kernel_t;
>         class fifo_file read;
> }
> 
> require {
>         type kernel_t;
>         type unconfined_t;
>         class fifo_file write;
> }
> 
> #============= kernel_t ==============
> allow kernel_t unconfined_t:fifo_file read;
> allow kernel_t unconfined_t:fifo_file write;
> ```

Does the custom selinux module need to be installed on the source qube or the destination qube? Or both?

**Comment by kalbasit** [comment]
> > ```cil
> > module fix-qvm-mount-folder 1.0;
> > 
> > require {
> >         type unconfined_t;
> >         type kernel_t;
> >         class fifo_file read;
> > }
> > 
> > require {
> >         type kernel_t;
> >         type unconfined_t;
> >         class fifo_file write;
> > }
> > 
> > #============= kernel_t ==============
> > allow kernel_t unconfined_t:fifo_file read;
> > allow kernel_t unconfined_t:fifo_file write;
> > ```
> 
> Does the custom selinux module need to be installed on the source qube or the destination qube? Or both?

The source qube, as in the one running the `qvm-mount-folder` command.
